management-node/server.go

```diff
@@ -0,0 +1,530 @@
+	}
+
+	wnl := pb.WorkerNodeList{}
+
+	for _, item := range gr.Kvs {
+		var wn wrpc.WorkerRPCClient
+		err := json.Unmarshal([]byte(item.Value), &wn)
+		if err != nil {
+			log.Error(err.Error())
+			return nil, err
+		}
+		wnl.Nodes = append(wnl.Nodes, wn.WN)
+	}
+
+	return &wnl, nil
+}
+
+// GetTaskList is gRPC that is called by schedctl with parameters
+// of DummyReq (for now). In future can be used for proper querying
+// and returns list of tasks that reside in db
+// TBD: can be done with pagination and limiting on rsp
+//		or even changed to stream
+func (md *ManagementNode) GetTaskList(ctx context.Context, req *pb.DummyReq) (*pb.TaskList, error) {
+
+	log.Info("GetTaskList")
+	opts := []clientv3.OpOption{
+		clientv3.WithPrefix(),
+	}
+
+	eCtx, cancel := context.WithTimeout(context.Background(),
+		time.Duration(Config.DialTimeout)*time.Second)
+	gr, err := md.etcd.Get(eCtx, EtcdTaskPrefix, opts...)
+	cancel()
+	if err != nil {
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	tl := pb.TaskList{}
+
+	for _, item := range gr.Kvs {
+		var task pb.Task
+		err := json.Unmarshal([]byte(item.Value), &task)
+		if err != nil {
+			log.Error(err.Error())
+			return nil, err
+		}
+		tl.Tasks = append(tl.Tasks, &task)
+	}
+
+	return &tl, nil
+}
+
+// Schedule is gRPC that is called by schedctl to schedule tasks
+// among connected workers. As parameters it receives list of tasks.
+// As return value it return list of tasks and their intermediate state.
+// Task call flow is following: received by Scheduled method ->
+// stored in db with pending state -> send to worker -> rxed confirmation ->
+// change state to scheduled , start dead timer -> store in db -> return
+// TBD: function is too big, split on smaller parts
+func (md *ManagementNode) Schedule(ctx context.Context, req *pb.TaskList) (*pb.TaskList, error) {
+	md.workerNodePoolMtx.RLock()
+	defer md.workerNodePoolMtx.RUnlock()
+
+	//make slice of connected workers
+	var workers []string
+	for k, _ := range md.workerNodePool {
+		workers = append(workers, k)
+	}
+
+	// if no connected workers set state of tasks requested to schedule to
+	// dead
+	// TBD: make it in waiting for active workers to be scheduled
+	if len(workers) == 0 {
+
+		log.Warning("There are no connected workers: all tasks are dead")
+		for _, task := range req.Tasks {
+			task.State = common.TaskStateDead
+			task.Id = uuid.New().String()
+		}
+
+		if err := md.setTasksToDb(req.Tasks); err != nil {
+			log.Error(err.Error())
+			return nil, err
+		}
+
+		return req, nil
+	}
+
+	// msg type for internal use
+	type msg struct {
+		err  error
+		task *pb.Task
+	}
+
+	// channel for connecting onRsp handlers
+	// to the current go routine
+	messanger := make(chan msg, 1)
+
+	for i, task := range req.Tasks {
+
+		// all the ids are generated by management node
+		id := uuid.New().String()
+		workerId := workers[0]
+
+		// RR worker id calculation
+		if Config.SchedulerAlgo == SchedAlgoRR {
+			workerId = workers[int(math.
```

Comment:
If you schedule tasks one-by-one, then using this algorithm it looks like they all will be scheduled on the first worker node? Unless I'm missing something.

Reply:
i is index going from 0 to len(workers)
let's assume len(workers) = 10
math.Mod(1,10) = 1, math.Mod(2,10) = 2 .... math.Mod(10,10) = 0 , math.Mod(11,10) = 1